src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,376 @@
+                }
+                else if (initialEditKind == SemanticTokenEditKind.Insert)
+                {
+                    orderedEditNumber = AddInsertionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        groupedSemanticToken: newGroupedSemanticTokens[orderedTokenNumber],
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+                else
+                {
+                    Contract.ThrowIfFalse(initialEditKind == SemanticTokenEditKind.Delete);
+                    orderedEditNumber = AddDeletionEdit(
+                        indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
+
+            // Local functions
+            static int AddUpdateEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int orderedEditNumber,
+                SemanticToken groupedSemanticToken,
+                int editStartPosition)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                groupedSemanticToken.AddToEnd(tokensToInsert);
+
+                // For simplicitly, we only allow an "update" (i.e. a dual insertion/deletion) to be
+                // combined with other updates.
+                var updatedOrderedEditNumber = GetUpdatedOrderedEditNumber(
+                    SemanticTokenEditKind.Update, indexToEditKinds, editIndices, orderedEditNumber);
+
+                var deleteCount = 5 + 5 * (updatedOrderedEditNumber - orderedEditNumber);
+                for (var i = 1; i <= updatedOrderedEditNumber - orderedEditNumber; i++)
+                {
+                    newGroupedSemanticTokens[editIndices[orderedEditNumber + i]].AddToEnd(tokensToInsert);
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: editStartPosition, deleteCount: deleteCount, data: tokensToInsert.ToArray()));
+                return updatedOrderedEditNumber;
+            }
+
+            static int AddInsertionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int orderedEditNumber,
+                SemanticToken groupedSemanticToken,
+                int editStartPosition)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                groupedSemanticToken.AddToEnd(tokensToInsert);
```

Comment:
same question about not passing this in, and just having the below loop start at 0.